posts/2017-10-19-3rd-party-rest-api-to-graphql-serverless.md

```diff
@@ -0,0 +1,551 @@
+    }
+    return perPage
+  }
+  if (!!first) {
+    const perPage = first
+    const start = 1
+    const skip = 0
+
+    return { start, perPage, skip }
+  }
+  if (!!last && !!total) {
+    const cursor = total - last
+    const perPage = minPerPage(total, last)
+    const start = Math.ceil(cursor / perPage) || 1
+    const skip = cursor % perPage
+
+    return { start, perPage, skip }
+  }
+  if (!!count) {
+    const perPage = offset > count ? minPerPage(offset, count) : count
+    const start = offset > count ? Math.ceil(offset / perPage) : 1
+    const skip = perPage - count <= 0 ? 0 : perPage - count
+
+    return { start, perPage, skip }
+  }
+  return {}
+}
+```
+
+It's used as part of the resolver's arguments in the following manner:
+
+```javascript
+//...
+fetchUserPhotos({ flickr, userId, ...pagination({...args, total}) })
+//...
+```
+
+Pagination only works in one of three ways at present: by the `first` number of results, the `last` number of results from the total, or a `count` that is a subsection of the results offset by a given number from the start of the result set.
+
+### And That's It!
+
+That about covers all the basic project components.
+
+There are some other utility functions I didn't cover here, mostly error handling and validation related. You'll have to dig through the [source code](https://github.com/Saeris/Flickr-Wormhole) if you want to learn how they work. ;)
+
+## Room for Improvement
+
+There are a few things this project is still missing. You may have even picked up on them already.
+
+Here's a short list:
+  - Right now the gateway only supports queries. There's no way to mutate data, meaning you can't do things like add comments or upload new photos to Flickr using the gateway. I didn't build out that functionality because I don't need it for my website as of yet.
+  - Because of the above, requests that require authentication are also unsupported at this time. These include certain read methods (e.g. fetching a user's galleries or a private list of favorites).
+  - I would like to add a shared cache between gateway instances using Amazon's Elasticache service. That way it could be possible to implement cursor based pagination and temporary caching of more sensitive data, which would require token based authentication and session tracking.
+  - Pagination sucks, a lot. It's got the most room for improvement.
+  - I'm working on adding more logging and monitoring tools, such as additional transports for [Winston](https://github.com/winstonjs/winston) to services like [Loggly](https://www.loggly.com/) or [Sentry](https://sentry.io/), and [Apollo Optic](https://www.apollodata.com/optics/)'s new [Engine](https://dev-blog.apollodata.com/apollo-engine-and-graphql-error-tracking-e7dd3ce8b99d) platform, which requires a Docker Container running on something like [Amazon's EC2 Container Service](https://aws.amazon.com/ecs/). These would help in tuning the performance of the gateway and monitoring for errors.
```

Comment:
Change Optic's to Optics' (product name was misspelled/mistaken for a grammar error)